### ISSUE TITLE
Wired rendering results

### ISSUE THREAD

**Issue by Vilour** [graphdeco-inria/gaussian-splatting#31]
If we move our views up and down, the color pattern will become strange. We could notice there is a plane seems like cut the whole scene half. The color becomes normal when our camera pose was reset to training view poses.
![动画1-min](https://github.com/graphdeco-inria/gaussian-splatting/assets/22342670/5b582b41-d24a-40b0-8fde-367049be3c69)
This phenomenon happens globally. This plane is located at certain height.
![动画2-min](https://github.com/graphdeco-inria/gaussian-splatting/assets/22342670/1419ce73-cfc2-4120-a41b-16001fe913d5)
Besides, we noticed when rendering the initial point clouds, some points are disappeared when the camera moves up and down.
![动画4-min](https://github.com/graphdeco-inria/gaussian-splatting/assets/22342670/fccee96e-2e4c-4615-b843-44737c3d74a6)
I am wandering if the issues were connected to raster? Are they actually the same problems?

**Comment by grgkopanas** [comment]
@Snosixtyboo maybe a culling issue?

**Comment by Snosixtyboo** [comment]
That's really very strange. I don't think I've seen this before. Also it doesn't really make sense that it occurs on both the initial point cloud and the trained model...

@Vilour would it be possible to get the trained model directory so we can take a look? We don't need the images if they are proprietary.

**Comment by Vilour** [comment]
@grgkopanas @Snosixtyboo May I have your email addresses? It could be an issue if the dataset went public. I could send you models privately.

**Comment by Snosixtyboo** [comment]
Sure, it's no secret: kerbl@cg.tuwien.ac.at

**Comment by Snosixtyboo** [comment]
Ok, I had time to take a look. I didn't find a clear indication of what did not work out, but I have some suggestions:
 
- Depth sorting was disabled for the initial point cloud rendering. I fixed that and pushed the code (Windows binaries soon). The points were always there, but they were not drawn in order.
- It seems there are some different requirements with regards to lighting in the scene. The same points on the same pillar need to model two different illuminations (see below). Trying to model this may cause unexpected behavior. This may be avoided by a fixed exposure and no auto-white balance, if those were enabled during the capture. ![image](https://github.com/graphdeco-inria/gaussian-splatting/assets/40643808/bd712a71-e6bb-43b0-8906-740a87812922) ![image](https://github.com/graphdeco-inria/gaussian-splatting/assets/40643808/c6876c52-32a3-45ad-8545-c3e401e775ff)
- The cameras around index 404 seem like there was an error calibrating their position? The renderings don't look good and they are the only ones that seem to be **upside down** 
![image](https://github.com/graphdeco-inria/gaussian-splatting/assets/40643808/554b2b74-6f61-432f-af13-709d059ebbcf)
- Generally, the initial point cloud is very dense and the scene is rather large. I would highly recommend for such a setup to reduce the position and scaling learning rates (see the FAQ on extensive data sets). This should yield higher-quality results. 

I hope that some of these steps will help you!

**Comment by Snosixtyboo** [comment]
My current impression is that it has mainly to do with different exposure/white balance in the photographs. The same area seen from the same direction (but different proximities) appears to be modeled with two very different levels of brightness. I don't have the original images, but I assume they have these properties too. Our method and related ones can encode view-dependent effects, but only if they are direction independent. 

| ![image](https://github.com/graphdeco-inria/gaussian-splatting/assets/40643808/8ccb601d-6bf5-4bd9-aefb-68bde8cdc946) | ![image](https://github.com/graphdeco-inria/gaussian-splatting/assets/40643808/a20233bd-603c-40f6-aefb-73a82b0aa0c5)|
| --- | ---

If the appearance of regions changes even when looking at them from the same direction (but at different proximities) the model is forced to use tricks.

**Comment by Vilour** [comment]
Thank you for your careful investigation! I tried to investigate the camera poses and found that some were clearly wrong. I removed them and trained again, but the phenomenon still existed. It is true that the exposure varied in the sequence. 
![1682231168 272000_F](https://github.com/graphdeco-inria/gaussian-splatting/assets/22342670/34faec8c-e493-4a55-867c-e40207603dad)
![1682231265 397000_B](https://github.com/graphdeco-inria/gaussian-splatting/assets/22342670/07ce9cbb-3ff6-44b3-87ea-75f2560fbcce)
The same pillar has different colors in different images. I'm wondering that if the varying rendering color could relate to the spherical harmonics? Our sequence is captured by a hand-held device, and a person walked around to capture the whole scene. As a result, the trajectory of the cameras is distributed on a plane. The variation of z axis (gravity direction) of the poses is little, which causes the spherical harmonics failed to generalize with different directions. Along with the varying exposure of images, the wired rendering results appeared. Is it possible?

**Comment by Vilour** [comment]
By the way, I have captured a new sequence and the exposure was set constant. The weird line still existed. 
![动画4](https://github.com/graphdeco-inria/gaussian-splatting/assets/22342670/79a6d1a2-6dc6-49cf-b697-5e5d22364533)
I have uploaded the data onto [Google driver](https://drive.google.com/drive/folders/1ZYc0wlLVFa2R4VaM4Swp_FXDD84tlam_?usp=sharing). I hope this could be helpful.

**Comment by grgkopanas** [comment]
Hi,

indeed spherical harmonics are not great in extrapolation, we have observed similar issues and we tried to improved the quality by first training the DC component of the spherical harmonics and then introducing the extra bands every 1000 iterations. You can try disabling the SH completly to see if this improves the results by using this argument:

https://github.com/graphdeco-inria/gaussian-splatting/blob/main/arguments/__init__.py#L49

**Comment by Vilour** [comment]
Thank you for your suggestion! The phenomenon disappeared when I reduced the SH to 2! It works! I rendered the test poses using render.py script. However, when I tried to render the model by SIBR_viewers, the pre-built windows apps only rendered black images. The SIBR viewer could not be compatible with models whose SH is smaller than 3.

**Comment by Snosixtyboo** [comment]
> Thank you for your suggestion! The phenomenon disappeared when I reduced the SH to 2! It works! I rendered the test poses using render.py script. However, when I tried to render the model by SIBR_viewers, the pre-built windows apps only rendered black images. The SIBR viewer could not be compatible with models whose SH is smaller than 3.

Great to hear!

Yes that makes a lot of sense now that i think about it, I never accounted for that. Could you maybe kindly upload the model somewhere once more, and I'll modify the viewer so that it can load your models?

Best,
Bernhard

**Comment by Snosixtyboo** [comment]
@Vilour thanks a lot, I pushed new SIBR submodules and also updated the executables, they should now properly load models with lower SH degrees. Hth!

Btw, I still think given how dense your initial point cloud is and the amount of photos you have, if you lower the position/scaling learning rate (```position_lr_init```,```position_lr_final```, ```scaling_lr```, divide them by 3 or 10), you should get higher-quality results.

Best,
Bernhard

**Comment by hecodeit** [comment]
> > @Vilour thanks a lot, I pushed new SIBR submodules and also updated the executables, they should now properly load models with lower SH degrees. Hth!
> > Btw, I still think given how dense your initial point cloud is and the amount of photos you have, if you lower the position/scaling learning rate (`position_lr_init`,`position_lr_final`, `scaling_lr`, divide them by 3 or 10), you should get higher-quality results.
> > Best, Bernhard
> 
> Thanks a lot! The viewer helps a lot to investigate training results! Actually in the trained model I sent to you, the `scaling_lr` and `position_lr_init` have been down scaled by 100. However, in the viewer we could still observe the very large Gaussians.
> 
> <img alt="" height="260" src="https://user-images.githubusercontent.com/22342670/254905531-02e783b3-1fe2-44e2-b2ed-3204f6034895.jpg"><img alt="" height="260" src="https://user-images.githubusercontent.com/22342670/254905541-65b9f198-91b8-4cb7-b16f-6dd8cea8d4b8.jpg"><img alt="" height="260" src="https://user-images.githubusercontent.com/22342670/254905547-86b640de-9b8c-4be7-b750-bace21148259.jpg">
> In the paper has mentioned the large Gaussians tend to happen in the view dependent regions, and add some regularization terms will help with the popping artifacts. Since the our initial point cloud is quit dense, I'm wondering if we can add some limitations on the Gaussians to prevent it getting larger. For example, if we add some hyperparameters like `scaling_lr_final`, `scaling_lr_max_steps` or `scaling_lr_delay_mult`, will the popping artifacts be reduced?


@Vilour It may not related with your post: I'm having the texture-less surface problem, and want to know how do you generate point cloud like that? As I know colmap will "ignore texture-less surface".
